code/pngutils/pngutils.cpp

```diff
@@ -206,3 +210,557 @@ int png_read_bitmap(const char *real_filename, ubyte *image_data, ubyte *bpp, in
+			_apng_failed("couldn't finish fdat apng frame");
+		}
+
+		if (_dispose_op == 2) {
+			 // revert to previous; so save for later
+			 _frame_next.data = frame.data;
+		}
+		_compose_frame();
+		_frames.push_back(frame);
+	}
+	else {
+		nprintf(("apng", "apng next_frame; used old (%03lu) (%03i)\n", _frames.size(), current_frame));
+	}
+
+	if (current_frame < nframes-1) {
+		frame = _frames.at(current_frame++);
+	}
+}
+
+/*
+ * @brief return image size in bytes
+ */
+size_t apng_ani::imgsize()
+{
+	return _image_size;
+}
+
+
+/*
+ * @brief Get info about the apng
+ * @note Also validates the apng & sets it up to have frames read
+ *
+ * @retval PNG_ERROR_NONE (0), otherwise will raise exception
+ */
+int apng_ani::load_header()
+{
+	char filename[MAX_FILENAME_LEN];
+
+	strcpy_s(filename, _filename.c_str());
+	char *p = strchr( filename, '.' );
+	if ( p != nullptr ) *p = 0;
+	strcat_s( filename, ".png" );
+
+	_cfp = cfopen( filename , "rb" );
+
+	if ( _cfp == nullptr) {
+		_apng_failed("couldn't open filename");
+	}
+
+	_reading = true;
+
+	ubyte sig[8];
+	if (cfread(sig, 8, 1, _cfp) != 1)  {
+		_apng_failed("cfread of png signature failed");
+	}
+	if (png_sig_cmp(sig, 0, 8) != 0) {
+		_apng_failed("file has invalid png signature");
+	}
+
+	_id = _read_chunk(_chunk_IHDR);
+
+	if (_id != id_IHDR || _chunk_IHDR.size != 25) {
+		_apng_failed("failed to read IHDR chunk");
+	}
+
+	w = png_get_uint_32(&_chunk_IHDR.data[8]);
+	h = png_get_uint_32(&_chunk_IHDR.data[12]);
+	_row_len = w * 4;
+	bpp = 32;  // we'll force all our frames to use this
+
+	// setup frames & keep bm_create happy
+	_image_size = _row_len * h;
+	frame.data.reserve(_image_size); // alloc only once
+	frame.data.assign(_image_size, 0); // all transparent black per spec
+	frame.rows.resize(h);
+	_frame_raw.data.resize(_image_size);
+	_frame_raw.rows.resize(h);
+	_frame_next.data.resize(_image_size);
+	_frame_next.rows.resize(h);
+	for (uint i = 0; i < h; ++i) {
+		frame.rows.at(i) =       &frame.data.at(i * _row_len);
+		_frame_raw.rows.at(i) =  &_frame_raw.data.at(i * _row_len);
+		_frame_next.rows.at(i) = &_frame_next.data.at(i * _row_len);
+	}
+
+	// read all data
+	while (!cfeof(_cfp)) {
+		_process_chunk();
+	}
+
+	// should be at EOF; attach to _frame_offsets to make next_frame code simpler
+	Assertion(cfeof(_cfp) != 0, "apng not at EOF, get a coder!");
+	_frame_offsets.push_back(cftell(_cfp));
+
+	// sanity checks
+	if (anim_time <= 0.0f) {
+		_apng_failed("animation duration <= 0.0f, bad data?");
+	}
+
+	if (nframes < 1) {
+		_apng_failed("animation didn't have any frames, is this a static png?");
+	}
+
+	// reset _cfp so it can be used for next frame
+	_reading = false;
+	if (cfseek(_cfp, _frame_offsets.at(0), CF_SEEK_SET) != 0) {
+		_apng_failed("couldn't seek to 1st fcTL offset");
+	}
+
+	return PNG_ERROR_NONE;
+}
+
+/*
+ * @brief cleanup resources
+ *
+ */
+void apng_ani::_cleanup_resources()
+{
+	png_destroy_read_struct(&_pngp, &_infop, nullptr);
+	cfclose(_cfp);
```

Comment:
good point, I'll add a check for ```nullptr``` before calling ```cfclose```.

Note that png_destroy_read_struct internally checks for and handles ```nullptr``` so no need for an additional check here.